### PR TITLE
Task 6/basic input

### PR DESCRIPTION
**Pull request by MrDmitry**

### CODE REVIEW

File: UnicornEngine/system/include/unicorn/system/Manager.hpp [hunk 1]
```diff
@@ -192,6 +201,9 @@ class Manager
      */
     void SetWindowCreationHint(WindowHint hint, int32_t value) const;
 
+    /** @brief  Polls all connected gamepads for new values */
+    void PollGamepads();
```
Comment (MrDmitry) [inline UnicornEngine/system/include/unicorn/system/Manager.hpp]: Also add a method to access the connected gamepads

File: UnicornEngine/system/source/Manager.cpp [hunk 2]
```diff
@@ -89,22 +105,42 @@ void Manager::Init()
...
 
-    MonitorCreated.emit();
+    GAMEPAD_ADAPTER::GamepadStateChanged.connect(this, &Manager::OnGamepadStateChanged);
+
+    {
+        std::vector<void*> gamepads = GAMEPAD_ADAPTER::GetConnectedGamepads();
+
+        input::Gamepad* pGamepad = nullptr;
+
+        for (std::vector<void*>::const_iterator cit = gamepads.cbegin(); cit != gamepads.cend(); ++cit)
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/Manager.cpp]: `for (auto const & g : gamepads)`

File: UnicornEngine/system/source/Manager.cpp [hunk 3]
```diff
@@ -89,22 +105,42 @@ void Manager::Init()
...
+        {
+            pGamepad = new input::Gamepad(*cit);
+            m_gamepads.insert( std::pair<uint32_t, input::Gamepad*>(pGamepad->GetId(), pGamepad) );
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/Manager.cpp]: emplace?

File: UnicornEngine/system/source/Manager.cpp [hunk 4]
```diff
@@ -169,6 +205,43 @@ void Manager::OnMonitorStateChanged(void* handle, MonitorMemento::State state)
...
+
+    switch (state)
+    {
+        case input::GamepadState::Connected:
+        {
+            if (cit == m_gamepads.cend())
+            {
+                input::Gamepad* pGamepad = new input::Gamepad(handle);
+
+                m_gamepads.insert(std::pair<uint32_t, input::Gamepad*>(pGamepad->GetId(), pGamepad));
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/Manager.cpp]: emplace?

File: UnicornEngine/system/source/Manager.cpp [hunk 5]
```diff
@@ -169,6 +205,43 @@ void Manager::OnMonitorStateChanged(void* handle, MonitorMemento::State state)
     MonitorCreated.emit();
 }
 
+void Manager::OnGamepadStateChanged(void* handle, input::GamepadState state)
+{
+    uint32_t id = static_cast<uint32_t>(reinterpret_cast<uintptr_t>(handle));
+    std::map<uint32_t, input::Gamepad*>::const_iterator cit = m_gamepads.find(id);
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/Manager.cpp]: auto?

File: UnicornEngine/system/source/Manager.cpp [hunk 6]
```diff
@@ -209,5 +282,13 @@ void Manager::SetWindowCreationHint(WindowHint hint, int32_t value) const
     WINDOW_MANAGER_ADAPTER::SetWindowCreationHint(hint, value);
 }
 
+void Manager::PollGamepads()
+{
+    for (std::map<uint32_t, input::Gamepad*>::const_iterator cit = m_gamepads.cbegin(); cit != m_gamepads.cend(); ++cit)
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/Manager.cpp]: Range based for?

File: UnicornEngine/system/source/Manager.cpp [hunk 7]
```diff
@@ -63,6 +66,19 @@ Window* Manager::GetWindow(uint32_t id) const
     return cit != m_windows.cend() ? cit->second : nullptr;
 }
 
+Window* Manager::GetFocusedWindow() const
+{
+    for (std::map<uint32_t, Window*>::const_iterator cit = m_windows.cbegin(); cit != m_windows.cend(); ++cit)
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/Manager.cpp]: std::find_if ?

File: UnicornEngine/system/include/unicorn/system/Manager.hpp [hunk 8]
```diff
@@ -192,6 +208,12 @@ class Manager
      */
     void SetWindowCreationHint(WindowHint hint, int32_t value) const;
 
+    /** @brief  Returns all connected gamepads */
+    const std::map<uint32_t, input::Gamepad*>& GetGamepads() const { return m_gamepads; }
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/include/unicorn/system/Manager.hpp:212]: noexcept?

File: UnicornEngine/system/include/unicorn/system/input/Gamepad.hpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    Gamepad& operator=(const Gamepad& other) = delete;
+
+    Gamepad(Gamepad&& other) = delete;
+    Gamepad& operator=(Gamepad&& other) = delete;
+
+    /** @brief  Destructs a gamepad object */
+    ~Gamepad();
+
+    /** @brief  Returns the id of the gamepad */
+    UNICORN_EXPORT uint32_t GetId() const { return m_id; }
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/include/unicorn/system/input/Gamepad.hpp:42]: noexcept for all trivial getters?

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp [hunk 10]
```diff
@@ -0,0 +1,1431 @@
...
+        Adapter::ConvertToUnicornModifiers(modifiers)
+    );
+}
+
+void unicornWindowFileDrop(GLFWwindow* handle, int count, const char** rawPaths)
+{
+    std::vector<std::string> paths;
+    paths.reserve(count);
+
+    for (int i = 0; i < count; ++i)
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/adapter/glfw/Adapter.cpp]: use std::transform ? or maybe just `std::vector<std::string> paths(rawPaths, rawPaths + count);`?

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp [hunk 11]
```diff
@@ -0,0 +1,1431 @@
...
+}
+
+std::vector<const char*> Adapter::GetRequiredVulkanExtensions()
+{
+    std::vector<const char*> extensions;
+
+    unsigned int glfwExtensionCount = 0;
+    const char** glfwExtensions = glfwGetRequiredInstanceExtensions(&glfwExtensionCount);
+
+    extensions.reserve(glfwExtensionCount);
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/adapter/glfw/Adapter.cpp]: `return std::vector<const char*>(glfwExtensions, glfwExtension + glfwExtensionCount);` ?

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp [hunk 12]
```diff
@@ -0,0 +1,1431 @@
...
+    return extensions;
+}
+
+std::vector<MonitorMemento> Adapter::GetMonitors()
+{
+    int count = 0;
+    GLFWmonitor** monitors = glfwGetMonitors(&count);
+
+    std::vector<MonitorMemento> result;
+    result.reserve(count);
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/adapter/glfw/Adapter.cpp:471]: `std::transform(monitors, monitors+count, result.begin(), [this](auto & m) { return this->GetMonitor(m); });`  or  something like that?

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp [hunk 13]
```diff
@@ -0,0 +1,1431 @@
...
+}
+
+std::vector<float> Adapter::GetGamepadAxes(void* handle)
+{
+    const int gamepad = static_cast<int>(reinterpret_cast<intptr_t>(handle));
+    int count = 0;
+
+    const float* axes = glfwGetJoystickAxes(gamepad, &count);
+
+    std::vector<float> result(count);
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/adapter/glfw/Adapter.cpp:600]: Constructor initialization and rvalue return?

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp [hunk 14]
```diff
@@ -0,0 +1,1431 @@
...
+    const int gamepad = static_cast<int>(reinterpret_cast<intptr_t>(handle));
+    int count = 0;
+
+    const unsigned char* buttons = glfwGetJoystickButtons(gamepad, &count);
+
+    std::vector<bool> result(count);
+
+    for (int i = 0; i < count; ++i)
+    {
+        result[i] = (GLFW_PRESS == buttons[i]) ? true : false;
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/adapter/glfw/Adapter.cpp:621]: std::transform?

File: UnicornEngine/system/source/input/Gamepad.cpp [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    bool anyUpdates = false;
+
+    {
+        std::vector<float> axes = GAMEPAD_ADAPTER::GetGamepadAxes(m_handle);
+
+        if (!m_axes.empty() && axes.size() != m_axes.size())
+        {
+            LOG_WARNING("Gamepad[%d]: number of axes changed during runtime", m_id);
+
+            m_axes.resize(axes.size());
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/input/Gamepad.cpp]: Remove double resize and leave only warning if. Resize won't do any action if sizes are equal, so you could just do it anyway out of the if scope, would not affect the performance and will make it a lot shorter.

File: UnicornEngine/system/source/input/Gamepad.cpp [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    {
+        std::vector<bool> buttons = GAMEPAD_ADAPTER::GetGamepadButtons(m_handle);
+
+        if (!m_buttons.empty() && buttons.size() != m_buttons.size())
+        {
+            LOG_WARNING("Gamepad[%d]: number of buttons changed during runtime", m_id);
+
+            m_buttons.resize(buttons.size());
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/input/Gamepad.cpp]: Remove double resize and leave only warning if. Resize won't do any action if sizes are equal, so you could just do it anyway out of the if scope, would not affect the performance and will make it a lot shorter.

File: UnicornEngine/system/source/input/Gamepad.cpp [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            LOG_WARNING("Gamepad[%d]: number of axes changed during runtime", m_id);
+
+            m_axes.resize(axes.size());
+        }
+        else if(m_axes.empty())
+        {
+            m_axes.resize(axes.size());
+        }
+
+        for (uint32_t i = 0; i < axes.size(); ++i)
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/input/Gamepad.cpp]: Could put it in a lambda and store as a functor, would make it possible to reuse it without copy paste.

File: UnicornEngine/system/source/input/Gamepad.cpp [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+            LOG_WARNING("Gamepad[%d]: number of buttons changed during runtime", m_id);
+
+            m_buttons.resize(buttons.size());
+        }
+        else if (m_buttons.empty())
+        {
+            m_buttons.resize(buttons.size());
+        }
+
+        for (uint32_t i = 0; i < buttons.size(); ++i)
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/input/Gamepad.cpp]: Could put it in a lambda and store as a functor, would make it possible to reuse it without copy paste.

File: demos/SanicJymper/main.cpp [hunk 19]
```diff
@@ -32,7 +154,14 @@ int main(int argc, char* argv[])
 
     if (unicornEngine->Init())
     {
-        unicorn::video::Graphics* pGraphics = unicornEngine->GetGraphics();
+        pGraphics = unicornEngine->GetGraphics();
+
+        const std::map<uint32_t, unicorn::system::input::Gamepad*>& gamepads = unicornEngine->GetGamepads();
```
Comment (ilia-glushchenko) [inline demos/SanicJymper/main.cpp]: `for (auto & g : unicornEngine->GetGamepads())` a lot shorter.

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp [hunk 20]
```diff
@@ -0,0 +1,1431 @@
...
+GammaRamp Adapter::GetGammaRamp(void* handle)
+{
+    const GLFWgammaramp* ramp = glfwGetGammaRamp(static_cast<GLFWmonitor*>(handle));
+
+    GammaRamp result = {};
+
+    if (ramp)
+    {
+        result.size = ramp->size;
+        result.red = new uint16_t[result.size];
```
Comment (ilia-glushchenko) [inline UnicornEngine/system/source/adapter/glfw/Adapter.cpp:540]: Maybe use unique_ptr and move it around so no one should remember to delete it?

File: UnicornEngine/system/source/Manager.cpp (hunk 7 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/Manager.cpp]: What would the benefit be?

File: UnicornEngine/system/source/Manager.cpp (hunk 2 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/Manager.cpp]: What would the benefit be?

File: UnicornEngine/system/source/Manager.cpp (hunk 5 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/Manager.cpp]: why?

File: UnicornEngine/system/source/Manager.cpp (hunk 6 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/Manager.cpp]: ~But it's not a vector~ herp derp for (meme : memes)

File: UnicornEngine/system/include/unicorn/system/Manager.hpp (hunk 8 above)
Comment (MrDmitry) [reply UnicornEngine/system/include/unicorn/system/Manager.hpp:212]: we're not using exceptions at all, why noexcept?

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp (hunk 12 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/adapter/glfw/Adapter.cpp:471]: what would the benefit be?

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp (hunk 20 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/adapter/glfw/Adapter.cpp:540]: don't quite see the need

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp (hunk 14 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/adapter/glfw/Adapter.cpp:621]: what would the benefit be?

File: UnicornEngine/system/source/input/Gamepad.cpp (hunk 17 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/input/Gamepad.cpp]: where is the copy paste?

File: UnicornEngine/system/source/input/Gamepad.cpp (hunk 18 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/input/Gamepad.cpp]: where is the copy paste?

File: demos/SanicJymper/main.cpp (hunk 19 above)
Comment (MrDmitry) [reply demos/SanicJymper/main.cpp]: I'm really not sold on the auto idea - to learn what the type is developer would have to open the header?

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp (hunk 12 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/adapter/glfw/Adapter.cpp:471]: not sold on this one at the moment

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp (hunk 20 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/adapter/glfw/Adapter.cpp:540]: Expanded documentation to cover the ability to point to the same buffer.

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp (hunk 13 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/adapter/glfw/Adapter.cpp:600]: Without std::transform it won't be possible due to conversion of values. And I am yet not sure about the std::transform

File: UnicornEngine/system/source/input/Gamepad.cpp (hunk 17 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/input/Gamepad.cpp]: gotcha

File: UnicornEngine/system/source/input/Gamepad.cpp (hunk 18 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/input/Gamepad.cpp]: gotcha

File: UnicornEngine/system/source/adapter/glfw/Adapter.cpp (hunk 14 above)
Comment (MrDmitry) [reply UnicornEngine/system/source/adapter/glfw/Adapter.cpp:621]: This method is called every frame and I would prefer to leave it as is until any problem with it occur. I could not find any credible materials regarding performance impacts of std::transform

**Review by krupitskas — APPROVED** [review]

**Pull request merged**